pkg/controller/servicemesh/webhook/controller.go

```diff
@@ -0,0 +1,212 @@
+	c, err := controller.New("webhook-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch secrets, but only where name is in secretNames
+	err = c.Watch(&source.Kind{Type: &corev1.Secret{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc: func(event event.CreateEvent) bool {
+			return contains(secretNames, event.Meta.GetName())
+		},
+		UpdateFunc: func(event event.UpdateEvent) bool {
+			return contains(secretNames, event.MetaNew.GetName())
+		},
+		// deletion and generic events don't interest us
+		DeleteFunc: func(event event.DeleteEvent) bool {
+			return false
+		},
+		GenericFunc: func(event event.GenericEvent) bool {
+			return false
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch MutatingWebhookConfigurations
+	err = c.Watch(&source.Kind{Type: &v1beta1.MutatingWebhookConfiguration{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(ns handler.MapObject) []reconcile.Request {
+			requests := []reconcile.Request{}
+			if strings.HasPrefix(ns.Meta.GetName(), injectorWebhookConfigNamePrefix) {
+				requests = append(requests, reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      secretNames[0],
+						Namespace: ns.Meta.GetName()[len(injectorWebhookConfigNamePrefix):],
+					},
+				})
+			}
+			return requests
+		})})
+	if err != nil {
+		return err
+	}
+
+	// Watch ValidatingWebhookConfigurations
+	err = c.Watch(&source.Kind{Type: &v1beta1.ValidatingWebhookConfiguration{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(ns handler.MapObject) []reconcile.Request {
+			requests := []reconcile.Request{}
+			if strings.HasPrefix(ns.Meta.GetName(), galleyWebhookConfigNamePrefix) {
+				requests = append(requests, reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      secretNames[1],
+						Namespace: ns.Meta.GetName()[len(galleyWebhookConfigNamePrefix):],
+					},
+				})
+			}
+			return requests
+		})})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// reconciles webhook configurations
+type reconciler struct {
+	common.ResourceManager
+	scheme *runtime.Scheme
+}
+
+// Reconcile reacts to changes in secrets by updating the corresponding webhook
+// configurations
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	// get secret
+	secret := &corev1.Secret{}
+	err := r.Client.Get(context.TODO(), request.NamespacedName, secret)
+	if err != nil {
+		// the secret has since disappeared. Ignore
+		return reconcile.Result{}, nil
+	}
+	caRoot, ok := secret.Data[rootCertKey]
+	if !ok {
+		logger.Error(errors.New("secret does not contain root certificate"), "Incomplete secret")
+		return reconcile.Result{}, nil
+	}
+
+	// is this for galley or the sidecar-injector?
+	if request.Name == secretNames[0] {
```

Comment:
Given these two blocks are almost identical, why not put those in a function?

Reply:
Because they operate on two different types with no common base except for `runtime.Object` - I guess I could use `unstructured` - but that would feel weird, exchanging type safety for brevity?

Reply:
I can remove a bit of the duplication though - the part that does the `Client.Update` below is pretty generic

Reply:
@jwendell I reduced duplication to a minimum without dropping the types. If you have an idea how to make it even more generic let me know!